lib/nvector_eigen.h

```diff
@@ -0,0 +1,528 @@
+#ifndef MEPBM_NVECTOR_EIGEN_H
+#define MEPBM_NVECTOR_EIGEN_H
+
+#include "sundials/sundials_nvector.h"
+#include "sundials/sundials_types.h"
+#include "eigen3/Eigen/Sparse"
+#include "eigen3/Eigen/Dense"
+#include <memory>
+#include <cmath>
+
+
+
+/**************************************************************************
+ * Functions whose memory locations are given to the N_Vector Ops structure
+ *************************************************************************/
+
+namespace NVectorOperations
+{
+  /// Returns the length of the vector
+  template <typename VectorType>
+  sunindextype
+  N_VGetLength( N_Vector v)
+  {
+    return (static_cast<VectorType*>(v->content))->size();
+  }
+
+
+
+  /// Creates a new N_Vector with the same ops field as an existing N_Vector. Allocates storage for the vector.
+  template <typename VectorType>
+  N_Vector
+  N_VClone( N_Vector w)
+  {
+    N_Vector v = N_VCloneEmpty(w);
+
+    // create memory for the vector in the new N_Vector
+    // the corresponding delete is called in N_VDestroy()
+    auto cloned = new VectorType(w->ops->nvgetlength(w));
+    v->content = cloned;
+    return v;
+  }
+
+
+
+  /// Creates a new N_Vector with the same ops field as an existing N_Vector. Does not allocate storage for the vector.
+  N_Vector
+  N_VCloneEmpty( N_Vector w)
+  {
+    N_Vector v = N_VNewEmpty();
+
+    // copy pointers to functions from existing vector
+    v->ops = w->ops;
+
+    return v;
+  }
+
+
+
+  /// Destroys the N_Vector and frees allocated memory
+  template<typename VectorType>
+  void
+  N_VDestroy( N_Vector v)
+  {
+    if (v->content != nullptr)
+    {
+      auto *content = static_cast<VectorType *>(v->content);
+      delete content;
+      v->content = nullptr;
+    }
+    N_VFreeEmpty(v);
+  }
+
+
+
+  /// Returns storage requirements for one N_Vector. This is a dummy function because it is not of interest.
+  void
+  N_VSpace(N_Vector v, sunindextype* lrw, sunindextype* liw)
+  {}
```

Comment:
a common way to deal with this is to do
```suggestion
  {
    assert (false);  // will terminate the program if ever actually called
  }
```

Reply:
I see why this is a way to do this. However, I think Sundials does call this function somewhere in CVODE. It just gives some diagnostic information that I don't think matters to me. So I'm pretty sure I can't actually terminate the program like this if this function is called.

Reply:
But then the implementation doesn't seem right to me. Some of the arguments appear to be *output* arguments, but by not setting them to anything you leave them undefined. If you don't have good estimates, say that your storage requirements are zero, or some other invalid or mildly useful value.

Reply:
In #120 I made the Space function give the length of the vector for the output asking about the number of `realtype` and zero for the output asking for the number of `int`. Does that implementation seem better to you?

Reply:
Yes, that seems reasonable.